Replace g_string_append_printf() with new pcmk__g_strcat() function

nrwahl2 opened the pull request:
This PR depends on #2879 and includes some overlapping commits. It introduces a new function `pcmk__g_strcat()` that's intended to replace `g_string_append_printf()` when the only interpolated objects are strings (i.e., when there are no integers (`%d`), pointers (`%p`), etc., to be appended to the string). This yields a performance improvement, with `crm_simulate` times decreasing or staying the same for almost every test input in `cts/scheduler/xml`.

Results below are compared to the "Use GString in pcmk__add_separated_word() commit", since it's assumed that that commit will be merged first via #2879.
```
# tools/pcmk_simtimes -S percent /tmp/after_pcmk__add_word_1000.txt /tmp/after_pcmk__g_strcat_1000.txt | tail -1
Averages: -0.11s  -2.62%
# tools/pcmk_simtimes -s 0.1 -S percent /tmp/after_pcmk__add_word_1000.txt /tmp/after_pcmk__g_strcat_1000.txt | tail -1
Averages: -0.40s  -3.48%
# tools/pcmk_simtimes -s 0.1 -p 3 -S percent /tmp/after_pcmk__add_word_1000.txt /tmp/after_pcmk__g_strcat_1000.txt | tail -1
Averages: -0.47s  -4.48%
```

clumens commented on include/crm/common/strings_internal.h:
```diff
@@ -142,6 +142,7 @@ bool pcmk__char_in_any_str(int ch, ...) G_GNUC_NULL_TERMINATED;
 int pcmk__strcmp(const char *s1, const char *s2, uint32_t flags);
 int pcmk__numeric_strcasecmp(const char *s1, const char *s2);
 void pcmk__str_update(char **str, const char *value);
+void pcmk__g_strcat(GString *buffer, ...);
```
If you add G_GNUC_NULL_TERMINATED to the function prototype, gcc will be able to warn on call sites that don't include a NULL at the end.  See pcmk__strcase_any_of, for instance.

nrwahl2 replied:
Oh neat. You know all the compiler tricks :)

clumens replied:
My answer to everything is "I bet the compiler can do this for me..."

nrwahl2 commented:
When I update, I'll change the first commit message from Refactor to Feature. I've been in refactor mode for a while...

clumens commented on lib/pengine/bundle.c:
```diff
@@ -176,11 +177,8 @@ create_docker_resource(pe_resource_t *parent, pe__bundle_variant_data_t *data,
-        int doffset = 0, dmax = 1024;
-        char *dbuffer = calloc(1, dmax+1);
+        GString *buffer = g_string_sized_new(4096);
+        GString *dbuffer = g_string_sized_new(1024);
```
These constants are used several places.  I can't imagine them changing in the future, but it might be nice to be ready just in case.

nrwahl2 replied:
Yeah, I don't have a strong preference. I had asked Ken about defining constants for GString sizes and he said they might vary from place to place and no reason to shoehorn all the calls into using one size.

You could make a decent case for defining constants per-file (rather than in a header). I don't expect it to make much difference and (like you) I don't expect to change them. We're unlikely to downsize them, and the nice thing about GStrings is that if they're too small, nothing catastrophic happens.

clumens commented on lib/cib/cib_attrs.c:
```diff
@@ -120,48 +106,45 @@ find_attr(cib_t *cib, const char *section, const char *node_uuid,
     }
 
     if (set_name) {
-        attr_snprintf(xpath_string, offset, XPATH_MAX, "//%s[@id='%.128s']", set_type,
```
You're getting rid of the `.128` here and several other places.  Is that alright?

nrwahl2 replied:
I presume so, since the goal is to build an XPath string to match something. My expectation is that the current version would simply fail to match its intended target, if our string doesn't fit into 128 characters. The `.128` was added in 4d978f5, and no rationale was given beyond ensuring that the string fits within the buffer.

clumens commented on lib/common/strings.c:
```diff
@@ -687,50 +687,46 @@ pcmk__str_table_dup(GHashTable *old_table)
- *                           and will be updated to new length of \p list
+ * \param[in,out] list       Pointer to current string list (may not be \p NULL)
+ * \param[in]     init_size  \p list will be initialized to this size, if it
+ *                           needs initialization
```
It's worth making it more clear that `init_size` can be 0.

nrwahl2 replied:
Fair point, I will

clumens commented on lib/common/strings.c:
```diff
@@ -1199,6 +1199,35 @@ pcmk__str_update(char **str, const char *value)
+    }
+    va_end(ap);
+}
+
```
A good function to add unit tests for (aren't they all).

nrwahl2 replied:
Agreed, slipped my mind :) I'll add them

nrwahl2 commented on lib/common/strings.c:
```diff
@@ -687,50 +687,46 @@ pcmk__str_table_dup(GHashTable *old_table)
+        if (init_size > 0) {
+            *list = g_string_sized_new(init_size);
+        } else {
+            *list = g_string_new(NULL);
```
Really this is redundant and only there for clarity. We could just use `g_string_sized_new(init_size)` no matter what. `g_string_new(NULL)` calls `g_string_sized_new(2)`, and `g_string_sized_new(size)` creates a string of size `MAX(size, 64)`. The GLib docs don't advertise that behavior though, so I consider it an implementation detail that we shouldn't assume.

nrwahl2 commented:
No changes; rebased on current main

nrwahl2 commented:
Huh, my unit tests must have gone into a different branch...

kgaillot commented on lib/common/options.c:
```diff
@@ -656,32 +656,48 @@ pcmk__format_option_metadata(const char *name, const char *desc_short,
-            }
+        // The standard requires a parameter type
+        if (opt_type == NULL) {
+            continue;
```
Since this is an internal function, this would be a bug, so maybe just assert that the \*desc\* and type fields are non-null so we're forced to come up with something reasonable rather than silently ignore a malformed entry

Though in that case, we should add regression tests for all callers so problems don't slip through

nrwahl2 replied:
The Makefile already runs `daemons/<daemon>/pacemaker-$daemon metadata > pacemaker-$daemon.xml` for each daemon that supports it, via `mk/man.mk`. So that makes sure we don't assert. If we want, we can add regression tests to `cts-cli` that actually check the output. I'll do that since it should be easy, and we can decide whether to keep it.

kgaillot commented on lib/common/options.c:
```diff
@@ -640,12 +639,11 @@ pcmk__format_option_metadata(const char *name, const char *desc_short,
-                              "  <version>%s</version>\n",
-                              name, PCMK_OCF_VERSION);
+    pcmk__g_strcat(s,
+                   "<?xml version=\"1.0\"?>"
```
This should end with a newline

nrwahl2 replied:
Good catch

nrwahl2 replied:
Oh, that one didn't end in a newline before the change either. I wondered about whether that was intentional or not. I'll still add one as long as it doesn't have any side effects.

kgaillot commented on lib/common/options.c:
```diff
@@ -640,12 +639,11 @@ pcmk__format_option_metadata(const char *name, const char *desc_short,
-                              name, PCMK_OCF_VERSION);
+    pcmk__g_strcat(s,
+                   "<?xml version=\"1.0\"?>"
+                   "<resource-agent name=\"", name, "\">\n"
```
It would make sense to add version= with the Pacemaker version

nrwahl2 replied:
Looks like this function is only called to build the metadata string for Pacemaker daemons (correct me if I'm wrong), so yeah that would make sense.

kgaillot commented on lib/common/options.c:
```diff
@@ -691,37 +689,37 @@ pcmk__format_option_metadata(const char *name, const char *desc_short,
-                                       "<content type=\"%s\">\n", opt_type);
+                pcmk__g_strcat(s, "      "
+                               "<content type=\"", opt_type, "\">\n", NULL);
             }
```
I think this would be easier to follow with

    ... <content type=...
    if (opt_default) != NULL) { ...default=... }
    ...>\n...

and we can avoid the repeated block below by doing that first regardless, then checking for select (just the close tag needs attention)

nrwahl2 replied:
Agreed

kgaillot reviewed (commented):
nice

nrwahl2 commented:
Updated to address review. This includes a new commit on top to add tests for daemon metadata, and a new commit in the middle that adds missing descriptions to two controller options (which caused an assertion when we added assertions for that in this PR).

kgaillot reviewed (approved):
Looks good. I forgot about the man page generation, that would have been sufficient, but no harm in having extra testing.
